src/transport/SecurePairingSession.cpp

```diff
@@ -218,68 +280,242 @@ CHIP_ERROR SecurePairingSession::AttachHeaderAndSend(uint8_t msgType, System::Pa
 {
     CHIP_ERROR err = CHIP_NO_ERROR;
 
-    VerifyOrExit(info != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
-    VerifyOrExit(info_len > 0, err = CHIP_ERROR_INVALID_ARGUMENT);
-    VerifyOrExit(mPairingComplete, err = CHIP_ERROR_INCORRECT_STATE);
+    // Response message processing
+    const uint8_t * resp = msg->Start();
+    size_t resplen       = msg->TotalLength();
+    uint32_t saltlen     = 0;
+    uint64_t iterCount   = 0;
 
-    err = session.InitFromSecret(mKe, mKeLen, nullptr, 0, info, info_len);
+    const PBKDFParamResponse * pResp = reinterpret_cast<const PBKDFParamResponse *>(resp);
+
+    ChipLogDetail(Ble, "Received PBKDF param response");
+
+    VerifyOrExit(resp != nullptr, err = CHIP_ERROR_MESSAGE_INCOMPLETE);
+    VerifyOrExit(resplen >= sizeof(PBKDFParamResponse), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    iterCount = chip::Encoding::LittleEndian::Get64(reinterpret_cast<const uint8_t *>(&pResp->mIterations));
+    saltlen   = chip::Encoding::LittleEndian::Get32(reinterpret_cast<const uint8_t *>(&pResp->mSaltLength));
+
+    VerifyOrExit(resplen == sizeof(PBKDFParamResponse) + saltlen, err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    // Specifications allow message to carry a uint64_t sized iteration count. Current APIs are
+    // limiting it to uint32_t. Let's make sure it'll fit the size limit.
+    VerifyOrExit(CanCastTo<uint32_t>(iterCount), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+
+    // Update commissioning hash with the received pbkdf2 param response
+    err = mCommissioningHash.AddData(resp, resplen);
+    SuccessOrExit(err);
+
+    err = SetupSpake2p(static_cast<uint32_t>(iterCount), pResp->mSalt, saltlen);
+    SuccessOrExit(err);
+
+    err = SendMsg1();
     SuccessOrExit(err);
 
 exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        SendErrorMsg(Spake2pErrorType::kUnexpected);
+    }
     return err;
 }
 
-CHIP_ERROR SecurePairingSession::HandleCompute_pA(const PacketHeader & header, const System::PacketBufferHandle & msg)
+CHIP_ERROR SecurePairingSession::SendMsg1()
+{
+    uint8_t X[kMAX_Point_Length];
+    size_t X_len = sizeof(X);
+    uint16_t data_len; // Will be the same as X_len in practice.
+
+    System::PacketBufferHandle msg_pA;
+
+    CHIP_ERROR err = mSpake2p.BeginProver(reinterpret_cast<const uint8_t *>(""), 0, reinterpret_cast<const uint8_t *>(""), 0,
+                                          &mWS[0][0], kSpake2p_WS_Length, &mWS[1][0], kSpake2p_WS_Length);
+    SuccessOrExit(err);
+
+    err = mSpake2p.ComputeRoundOne(X, &X_len);
+    SuccessOrExit(err);
+    VerifyOrExit(CanCastTo<uint16_t>(X_len), err = CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    data_len = static_cast<uint16_t>(X_len);
+
+    msg_pA = System::PacketBuffer::NewWithAvailableSize(data_len);
+    VerifyOrExit(!msg_pA.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
+
+    {
+        BufBound bbuf(msg_pA->Start(), data_len);
+        bbuf.Put(&X[0], X_len);
+        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
```

Comment:
How could this fail, since the same length is passed?  Again, seems like memcpy would be more appropriate.